stablehlo/conversions/tosa/transforms/TosaRescaleLegalizeToStablehlo.cpp

```diff
@@ -0,0 +1,194 @@
+  unsigned outputBitWidth = outputQType
+                                ? outputQType.getStorageTypeIntegralWidth()
+                                : outputEType.getIntOrFloatBitWidth();
+  int32_t minOutputValue =
+      APInt::getSignedMinValue(outputBitWidth).getSExtValue();
+  int32_t maxOutputValue =
+      APInt::getSignedMaxValue(outputBitWidth).getSExtValue();
+  auto i32MinAttr = DenseElementsAttr::get(i32Type, {minOutputValue});
+  Value outputMin =
+      rewriter.create<stablehlo::ConstantOp>(loc, i32Type, i32MinAttr);
+  auto i32MaxAttr = DenseElementsAttr::get(i32Type, {maxOutputValue});
+  Value outputMax =
+      rewriter.create<stablehlo::ConstantOp>(loc, i32Type, i32MaxAttr);
+
+  // convert to i64 tensors
+  Value multiplierI64 =
+      rewriter.create<stablehlo::ConvertOp>(loc, i64Type, multiplier);
+  Value shiftI64 = rewriter.create<stablehlo::ConvertOp>(loc, i64Type, shift);
+  Value inputZpI64 =
+      rewriter.create<stablehlo::ConvertOp>(loc, i64Type, inputZp);
+  Value outputZpI64 =
+      rewriter.create<stablehlo::ConvertOp>(loc, i64Type, outputZp);
+  Value inputI64 = rewriter.create<stablehlo::ConvertOp>(loc, i64Type, input);
+  Value outputMinI64 =
+      rewriter.create<stablehlo::ConvertOp>(loc, i64Type, outputMin);
+  Value outputMaxI64 =
+      rewriter.create<stablehlo::ConvertOp>(loc, i64Type, outputMax);
+
+  Value adjustedInput = rewriter.create<stablehlo::SubtractOp>(
+      loc, i64Type, inputI64, inputZpI64);
+  Value adjustedShift =
+      rewriter.create<stablehlo::SubtractOp>(loc, i64Type, shiftI64, onesI64);
+
+  Value round = rewriter.create<stablehlo::ShiftLeftOp>(loc, i64Type, onesI64,
+                                                        adjustedShift);
+
+  Value r1 = rewriter.create<stablehlo::MulOp>(loc, i64Type, adjustedInput,
+                                               multiplierI64);
+  Value r2 = rewriter.create<stablehlo::AddOp>(loc, i64Type, r1, round);
+  Value r3 = rewriter.create<stablehlo::ShiftRightArithmeticOp>(loc, i64Type,
+                                                                r2, shiftI64);
+  Value r4 = rewriter.create<stablehlo::AddOp>(loc, i64Type, r3, outputZpI64);
+  Value r5 = rewriter.create<stablehlo::ClampOp>(loc, i64Type, outputMinI64, r4,
+                                                 outputMaxI64);
+
+  Value result;
+  if (outputQType) {
+    auto storageType = outputType.clone(outputQType.getStorageType());
+    Value r6 = rewriter.create<stablehlo::ConvertOp>(loc, storageType, r5);
+    result = rewriter.create<stablehlo::BitcastConvertOp>(loc, outputType, r6);
+  } else {
+    result = rewriter.create<stablehlo::ConvertOp>(loc, outputType, r5);
+  }
+  rewriter.replaceOp(op, {result});
+
+  return success();
+}
+
+struct TosaRescaleLegalizeToStablehloPass
+    : impl::TosaRescaleLegalizeToStablehloPassBase<
+          TosaRescaleLegalizeToStablehloPass> {
+  void runOnOperation() final {
```

Comment:
Per https://mlir.llvm.org/docs/PassManagement/#initialization, let's separate the pass initialization from the actual execution. [e.g.](https://github.com/openxla/stablehlo/blob/604a1ecaa1d010113e45f7e69348ec23e3d977a1/stablehlo/conversions/tosa/transforms/StablehloLegalizeToTosa.cpp#L474)

Reply:
done